vms/platformvm/state/state.go

```diff
@@ -2220,3 +2340,173 @@ func parseStoredBlock(blkBytes []byte) (blocks.Block, choices.Status, bool, erro
 
 	return blkState.Blk, blkState.Status, true, nil
 }
+
+func (s *state) PruneAndIndex(lock sync.Locker, log logging.Logger) error {
+	lock.Lock()
+	shouldPrune, err := s.shouldPrune()
+	if err != nil {
+		lock.Unlock()
+		return fmt.Errorf(
+			"failed to check if the database should be pruned: %w",
+			err,
+		)
+	}
+	if !shouldPrune {
+		lock.Unlock()
+		return nil
+	}
+
+	// It is possible that new blocks are added after grabbing this iterator. New
+	// blocks are guaranteed to be accepted and height-indexed, so we don't need to
+	// check them.
+	blockIterator := s.blockDB.NewIterator()
+	// Releasing is done using a closure to ensure that updating blockIterator will
+	// result in having the most recent iterator released when executing the
+	// deferred function.
+	defer func() {
+		blockIterator.Release()
+	}()
+
+	// While we are pruning the disk, we disable caching of the data we are
+	// modifying. Caching is re-enabled when pruning finishes.
+	//
+	// Note: If an unexpected error occurs the caches are never re-enabled.
+	// That's fine as the node is going to be in an unhealthy state regardless.
+	oldBlockIDCache := s.blockIDCache
+	s.blockIDCache = &cache.Empty[uint64, ids.ID]{}
+	lock.Unlock()
+
+	log.Info("starting state pruning and indexing")
+
+	var (
+		startTime  = time.Now()
+		lastCommit = startTime
+		lastUpdate = startTime
+		numPruned  = 0
+		numIndexed = 0
+	)
+
+	for blockIterator.Next() {
```

Comment:
Where is the early exit if the node shuts down?

Reply:
The VM takes care of shutting down here: https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/vm.go#L315-L352

Reply:
I don't think we've really added any of these for any of our indexing - it'll exit when the DB errors due to being closed.. (or when the process exits 🤷)

Reply:
And to ask a question I think I know the answer to, partial pruning will be handled gracefully (and the pruned value will only be written once we've iterated over all elements)?

Reply:
Hmmmm this _should_ be the case. But I think we should probably check for `blockIterator.Error()` before trying to write the pruned value. Specifically, because it doesn't seem like the corruptabledb has any handling for iterators.